ark/phenotyping/cluster.py

```diff
@@ -0,0 +1,151 @@
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        x_mesh (numpy.ndarray):
+            The x coordinate matrix of the weights vectors
+        y_mesh (numpy.ndarray):
+            The y coordinate matrix of the weights vectors
+        t (int):
+            The current iteration index
+        num_iters (int):
+            The maximum number of iterations
+
+    Returns:
+        numpy.ndarray:
+            The updated weights matrix
+    """
+
+    # update learning rate with asymptotic decay
+    decay_lr = learning_rate / (1 + t / (num_iters / 2))
+
+    # update sigma with asymptotic decay
+    decay_sig = sigma / (1 + t / (num_iters / 2))
+
+    # return a Gaussian centered around the winning coordinates
+    d = 2 * np.pi * decay_sig * decay_sig
+    ax = np.exp(-np.power(x_mesh - x_mesh.T[winning_coords], 2) / d)
+    ay = np.exp(-np.power(y_mesh - y_mesh.T[winning_coords], 2) / d)
+    g = (ax * ay).T * decay_lr
+
+    # update the weights based on the Gaussian neighborhood
+    weights += np.einsum('ij, ijk->ijk', g, sample - weights)
+
+    return weights
+
+
+def train_flowsom(pixel_mat, x_neurons, y_neurons, num_iters,
+                  sigma=1.0, learning_rate=0.5, random_seed=0):
+    """Trains the SOM by iterating through the each data point and updating the params
+
+    Args:
+        pixel_mat (pandas.DataFrame):
+            A matrix with pixel-level channel information for non-zero pixels in img_xr
+        x_neurons (int):
+            The number of x neurons to use
+        y_neurons (int):
+            The number of y neurons to use
+        num_iters (int):
+            The maximum number of iterations for training
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        random_seed (int):
+            The seed to set for random weight initialization
+
+    Returns:
+        numpy.ndarray:
+            The weights matrix after training
+    """
+
+    # define the random generator
+    rand_gen = np.random.RandomState(random_seed)
+
+    # initialize the weights and normalize
+    weights = rand_gen.rand(x_neurons, y_neurons, pixel_mat.shape[1]) * 2 - 1
+    weights /= np.linalg.norm(weights, axis=-1, keepdims=True)
+
+    # define the activation map
+    activation_map = np.zeros((x_neurons, y_neurons))
+
+    # define meshgrid coords for the weights matrix, convert to float (yikes, memory...)
+    x_mesh, y_mesh = np.meshgrid(np.arange(x_neurons), np.arange(y_neurons))
+    x_mesh = x_mesh.astype(float)
+    y_mesh = y_mesh.astype(float)
+
+    # define the iterations iterable, this is WRONG in MiniSOM
+    iterations = np.arange(num_iters) % pixel_mat.shape[0]
```

Comment:
I think to begin with, we should define this the way it's defined in FlowSOM, where the number of passes through the dataset is specified, and each pass through the dataset looks at all of the samples. This will make it easier to compare to what the previous implementations have done before we start changing things

Reply:
@ngreenwald yeah, it was confusing the way MiniSOM defines iterations, I had to double-check with the guy who maintains the repo exactly what was going on. So as you mention, `num_iters` should now be the number of passes through the dataset, not just the number of random data points to select with replacement (a la MiniSOM's style).